webui/react/src/components/LogViewerTimestamp.tsx

```diff
@@ -109,167 +103,155 @@ const LogViewerTimestamp: React.FC<Props> = ({
+    listRef.current?.resetAfterIndex(0);
+    setIsLastReached(false);
+    setIsLoadingFirstBatch(true);
+  }, [ logsDispatch ]);
+
+  const fetchAndAppendLogs =
+    useCallback((direction: DIRECTIONS, filters: LogViewerTimestampFilter): AbortController => {
+      const canceler = new AbortController();
+      let fetchArgs = null;
+      let isPrepend = false;
+
+      if (direction === DIRECTIONS.TAILING) {
+        fetchArgs = onFetchLogBefore(filters, canceler);
+        isPrepend = true;
+      }
 
-  /*
-   * Detect log viewer resize events to trigger
-   * recalculation of measured log entries.
-   */
-  useLayoutEffect(() => {
-    const throttleFunc = throttle(DEFAULT_RESIZE_THROTTLE_TIME, () => {
-      if (!container.current) return;
-      setConfig(measureLogs(logs));
-    });
+      if (direction === DIRECTIONS.OLDEST) {
+        fetchArgs = onFetchLogAfter({
+          ...filters,
+          timestampAfter: filters.timestampAfter?.subtract(1, 'millisecond'),
+        }, canceler);
+        isPrepend = false;
+      }
 
-    throttleFunc();
-  }, [ logs, measureLogs, resize ]);
+      if (fetchArgs) {
+        let buffer: TrialLog[] = [];
+        consumeStream(
+          fetchArgs,
+          event => {
+            const logEntry = fetchToLogConverter(event);
+            direction === DIRECTIONS.OLDEST ? buffer.unshift(logEntry) : buffer.push(logEntry);
+          },
+        ).then(() => {
+          if (buffer.length < TAIL_SIZE) setIsLastReached(true);
 
-  /*
-   * Check if user scroll is on top/bottom.
-   */
-  useLayoutEffect(() => {
-    setIsScrollOn({
-      bottom: (
-        scroll.scrollHeight
-        - scroll.viewHeight
-        - scroll.scrollTop
-        < SCROLL_BOTTOM_THRESHOLD
-      ),
-      top: (
-        scroll.scrollTop < SCROLL_TOP_THRESHOLD
-      ),
-    });
-  }, [ scroll ]);
+          // prevent loading other logs after adding the loaded ones
+          setIsOnBottom(direction === DIRECTIONS.TAILING);
+          setIsOnTop(direction === DIRECTIONS.OLDEST);
 
-  /*
-   * Automatically scroll to log tail (if tailing).
-   */
-  useLayoutEffect(() => {
-    const element = container.current;
+          addLogs(buffer, isPrepend);
 
-    if (
-      !element
-      || !isScrollOn.bottom
-      || direction !== DIRECTIONS.TAILING
-    ) return;
+          setIsLoadingFirstBatch(false);
 
-    setTimeout(() => {
-      element.scrollTo({ top: element.scrollHeight });
-    });
-  }, [ container, direction, isScrollOn.bottom, logs ]);
+          buffer = [];
+        });
+      }
+
+      return canceler;
+    }, [ addLogs, fetchToLogConverter, onFetchLogAfter, onFetchLogBefore ]);
+
+  const getItemHeight = (index: number): number => {
+    const log = logs[index];
+    if (!log) {
+      return charMeasures.height;
+    }
+
+    const columnWidth = listMeasure.width - ICON_WIDTH - dateTimeWidth;
+    const messageWidth = log.message.length * charMeasures.width;
```

Comment:
**suggestion:** One caveat to this is that it will miss the case of when there is a newline within the message itself. I think I previously calculated it by splitting up the message based on the newlines, calculate how much height each part takes up and adding them up.

Reply:
Whoops, I thought about this and forgot it! Thanks!